contracts/deposit-contracts/contract/src/main.rs

```diff
@@ -0,0 +1,214 @@
+            .unwrap_or_revert();
+    last_processed_deposit_counter_value += 1u64;
+    storage::write(
+        last_processed_deposit_counter_uref,
+        last_processed_deposit_counter_value,
+    );
+}
+
+#[no_mangle]
+pub extern "C" fn withdrawal() {
+    sec_check(vec![SecurityBadge::Admin]);
+    let destination_purse: URef = runtime::get_named_arg(RUNTIME_ARG_DEST_PURSE);
+    let amount: U512 = runtime::get_named_arg(RUNTIME_ARG_AMOUNT);
+    let deposit_purse: URef = runtime::get_key(KAIROS_DEPOSIT_PURSE)
+        .unwrap_or_revert_with(ApiError::MissingKey)
+        .into_uref()
+        .unwrap_or_revert()
+        .with_access_rights(AccessRights::READ_ADD_WRITE);
+    system::transfer_from_purse_to_purse(deposit_purse, destination_purse, amount, None)
+        .unwrap_or_revert();
+}
+
+#[no_mangle]
+pub extern "C" fn change_security() {
+    sec_check(vec![SecurityBadge::Admin]);
+    let admin_list: Option<Vec<Key>> =
+        get_optional_named_arg_with_user_errors(ADMIN_LIST, DepositError::InvalidAdminList);
+    // construct a new admin list from runtime arg
+    let mut badge_map: BTreeMap<Key, SecurityBadge> = BTreeMap::new();
+    if let Some(admin_list) = admin_list {
+        for account_key in admin_list {
+            badge_map.insert(account_key, SecurityBadge::Admin);
+        }
+    }
+    // remove the caller from the admin list
+    let caller = get_immediate_caller().unwrap_or_revert();
+    badge_map.remove(&caller);
+    security::change_sec_badge(&badge_map);
+}
+
+#[no_mangle]
+pub extern "C" fn call() {
+    let caller = Key::from(runtime::get_caller());
+
+    let admin_list: Option<Vec<Key>> =
+        get_optional_named_arg_with_user_errors(ADMIN_LIST, DepositError::InvalidAdminList);
+
+    let entry_points = {
+        let mut entry_points = EntryPoints::new();
+        entry_points.add_entry_point(entry_points::init());
+        entry_points.add_entry_point(entry_points::get_purse());
+        entry_points.add_entry_point(entry_points::deposit());
+        entry_points.add_entry_point(entry_points::withdrawal());
+        entry_points.add_entry_point(entry_points::incr_last_processed_deposit_counter());
+        entry_points.add_entry_point(entry_points::change_security());
+        entry_points
+    };
+    let mut named_keys = NamedKeys::new();
+    let event_dict = storage::new_dictionary(KAIROS_DEPOSIT_EVENT_DICT).unwrap_or_revert();
+    named_keys.insert(KAIROS_DEPOSIT_EVENT_DICT.to_string(), event_dict.into());
+    let last_processed_deposit_counter = storage::new_uref(u64::from(0u8));
+
+    named_keys.insert(
+        KAIROS_LAST_PROCESSED_DEPOSIT_COUNTER.to_string(),
+        last_processed_deposit_counter.into(),
+    );
+
+    let most_recent_deposit_counter = storage::new_uref(u64::from(0u8));
+    named_keys.insert(
+        KAIROS_MOST_RECENT_DEPOSIT_COUNTER.to_string(),
+        most_recent_deposit_counter.into(),
+    );
+
+    let (contract_hash, _) = storage::new_contract(
```

Comment:
This allows to use Casper anti-feature of upgradeable contracts. What about creating locked contract?

Reply:
What would need to be changed to disable the upgradability?

Reply:
Done.

Reply:
@jonas089 It would be more helpful to get more context before conversation is resolved. Right now I have to go through every resolved conversation, then figure out what change you made.

From what I see in https://github.com/cspr-rad/kairos/pull/51/files#diff-53df1ae74d6a66c198d5f0fbabfcbda5d7525c472042d7b6260918827ebf5b28R211-R217 you disabled storing contract uref in current account. That's weird...

Please see `new_contract` [documentation](https://docs.rs/casper-contract/latest/casper_contract/contract_api/storage/fn.new_contract.html):

![image](https://github.com/cspr-rad/kairos/assets/6506780/b40ce20b-c4cf-4788-ba18-990f7f714ddd)

Reply:
I just fixed this. Sorry, I was mistaken.

Reply:
In the latest commit I use locked_contract and am adding the package to the installers named keys

Reply:
Thanks, I see it's fixed:

![image](https://github.com/cspr-rad/kairos/assets/6506780/39b88d5e-c099-4b54-849f-4c59c151ab18)